Adds tutorial for creating a worker (#9179)
PrefectHQ/prefect: docs/tutorials/creating-a-worker.md

---
description: Learn how to create a Prefect worker to run your flows.
tags:
    - work pools
    - workers
    - orchestration
    - flow runs
    - deployments
    - storage
    - infrastructure
    - tutorial
    - recipes
---

> # Creating a Worker

Reviewer: Any thoughts on this heading vs "Developing a worker"? I'm wondering if users who just want to start a worker or learn more about workers might end up here accidentally.

Reviewer 2: +1 this title is a bit too vague. What about "Developing a new worker type"?

    The work pool that a worker polls for flow runs has a [base job template](/concepts/work-pools/#base-job-template) associated with it which is the contract for how configuration values populate for each flow run.
    
    The keys in the `job_configuration` section of this base job template match the worker's configuration class attributes. The values in the `job_configuration` section of the base job template are used to populate the attributes of the worker's configuration class.

    The work pool creator gets to decide how they want to populate the values in the `job_configuration` section of the base job template. The values can be hard-coded, templated using placeholders, or a mix these two approaches. Because you, as the worker developer, don't know how the work pool creator will populate the values, so you should set sensible defaults for your configuration class attributes.

### Implementing a `BaseJobConfiguration` Subclass

A worker developer defines the configuration their worker needs to function with a class extending [`BaseJobConfiguration`](/api-ref/prefect/workers/base/#prefect.workers.base.BaseJobConfiguration). 

> `BaseJobConfiguration` has attributes that are common to all workers: 

Author: `type` isn't inferred from the class name and needs to be explicitly set right now.

| Attribute | Description |
| --------- | ----------- |
| `name` | The name to assign to the created execution environment. |
| `env` | Environment variables to set in the created execution environment. |
| `labels` | The labels assigned to the created execution environment for metadata purposes. |
| `command` | The command to use when starting a flow run. |

Prefect sets values for each attribute before giving the configuration to the worker. If you want to customize the values of these attributes, use the [`prepare_for_flow_run`](/api-ref/prefect/workers/base/#prefect.workers.base.BaseJobConfiguration.prepare_for_flow_run) method.

> Here's an example `prepare_for_flow_run` method that sets add a label to the execution environment:

Reviewer 3: ```suggestion
Here's an example `prepare_for_flow_run` method that adds a label to the execution environment:
```

Workers handle creating execution environments using provided configuration. Workers also observe the execution environment as the flow run executes and report any crashes to the Prefect API.

To implement a worker, you must implement the `BaseWorker` class and provide it with the following attributes:

| Attribute | Description | Required |
| --------- | ----------- | -------- |
| `type` | The type of the worker. | Yes |
| `job_configuration` | The configuration class for the worker. | Yes |
| `job_configuration_variables` | The template variables class for the worker. | No |
| `_documentation_url` | Link to documentation for the worker. | No |
| `_logo_url` | Link to a logo for the worker. | No |
| `_description` | A description of the worker. | No |

In addition to the attributes above, you must also implement a `run` method. The `run` method is called for each flow run the worker receives for execution from the work pool.

The `run` method has the following signature:

```python
 async def run(
        self, flow_run: FlowRun, configuration: BaseJobConfiguration, task_status: Optional[anyio.abc.TaskStatus] = None,
    ) -> BaseWorkerResult:
        ...
```

> The `run` method is passed the flow run to execute, the execution environment configuration for the flow run, and a task status object that allows the worker to track whether the flow run was submitted successfully.

Reviewer: ```suggestion
The `run` method is passed: the flow run to execute, the execution environment configuration for the flow run, and a task status object that allows the worker to track whether the flow run was submitted successfully.
```

The `run` method must also return a `BaseWorkerResult` object. The `BaseWorkerResult` object returned contains information about the flow run execution. For the most part, you can implement the `BaseWorkerResult` with no modifications like so:

```python
from prefect.workers.base import BaseWorkerResult

class MyWorkerResult(BaseWorkerResult):
    """Result returned by the MyWorker."""
```

If you would like to return more information about a flow run, then additional attributes can be added to the `BaseWorkerResult` class.

### Worker Implementation Example

> Below is an example of a worker implementation. This example is not intended to be a complete implementation but to illustrate the abovementioned concepts.

Reviewer: ```suggestion
Below is an example of a worker implementation. This example is not intended to be a complete implementation but to illustrate the aforementioned concepts.
```

            description="CPU allocation for the execution environment."
        )

class MyWorkerResult(BaseWorkerResult):
    """Result returned by the MyWorker."""

class MyWorker(BaseWorker):
    type = "my-worker"
    job_configuration = MyWorkerConfiguration
    job_configuration_variables = MyWorkerTemplateVariables
    _documentation_url = "https://example.com/docs"
    _logo_url = "https://example.com/logo"
    _description = "My worker description."

    async def run(
        self, flow_run: FlowRun, configuration: BaseJobConfiguration, task_status: Optional[anyio.abc.TaskStatus] = None,
    ) -> BaseWorkerResult:
        # Create the execution environment and start execution
        job = await self._create_and_start_job(configuration)

        if task_status:
            task_status.started(job.id) # Use a unique ID to mark the run as started

        # Monitor the execution
        job_status = await self._watch_job(job, configuration)

        exit_code = job_status.exit_code if job_status else -1 # Get result of execution for reporting
        return MyWorkerResult(
            status_code=exit_code,
            identifier=job.id,
        )
```

> Most of the execution logic is omitted from the example above, but it shows that the typical flow for execution in the `run` method is:

Reviewer: ```suggestion
Most of the execution logic is omitted from the example above, but it shows that the typical flow run for execution in the `run` method is:
```

    1. Create the execution environment and start the flow run execution
    2. Mark the flow run as started via the passed `task_status` object
    3. Monitor the execution
    4. Get the execution's final status from the infrastructure and return a `BaseWorkerResult` object

To see other examples of worker implementations, see the [`ProcessWorker`]((/api-ref/prefect/workers/process/)) and [`KubernetesWorker`](https://prefecthq.github.io/prefect-kubernetes/worker/) implementations.

### Integrating with the Prefect CLI

Workers can be started via the Prefect CLI by providing the `--type` option to the `prefect worker start` CLI command. To make your worker type available via the CLI, it must be available at import time. 

If your worker is in a package, you can add an entry point to your setup file in the following format:

```python
entry_points={
    "prefect.collections": [

>         "my_pacakge_name = my_worker_module",

Reviewer: ```suggestion
        "my_package_name = my_worker_module",
```
